pytorch_lightning/utilities/parsing.py

```diff
@@ -197,72 +197,70 @@ def __repr__(self):
         return out
 
 
-def lightning_hasattr(model, attribute):
-    """ Special hasattr for lightning. Checks for attribute in model namespace,
-        the old hparams namespace/dict, and the datamodule. """
+def lightning_get_all_attr_holders(model, attribute):
+    """ Special attribute finding for lightning.  Gets all of the objects or dicts that holds attribute.
+            Checks for attribute in model namespace, the old hparams namespace/dict, and the datamodule. """
     trainer = getattr(model, 'trainer', None)
 
-    attr = False
+    holders = []
+
     # Check if attribute in model
     if hasattr(model, attribute):
-        attr = True
+        holders.append(model)
+
     # Check if attribute in model.hparams, either namespace or dict
-    elif hasattr(model, 'hparams'):
-        if isinstance(model.hparams, dict):
-            attr = attribute in model.hparams
-        else:
-            attr = hasattr(model.hparams, attribute)
+    if hasattr(model, 'hparams'):
+        if attribute in model.hparams:
+            holders.append(model.hparams)
+
     # Check if the attribute in datamodule (datamodule gets registered in Trainer)
-    if not attr and trainer is not None:
-        attr = hasattr(trainer.datamodule, attribute)
+    if trainer is not None and trainer.datamodule is not None and hasattr(trainer.datamodule, attribute):
+        holders.append(trainer.datamodule)
 
-    return attr
+    return holders
+
+
+def lightning_get_first_attr_holder(model, attribute):
```

Comment:
do we really need this? is also not very consistent as the name says "first" but you take last as `[-1]`

Reply:
As mentioned in line 229, the last is taken to preserve backwards compatibility. In case somebody is relying on the attribute candidate order